model/project_ref.go

```diff
@@ -420,6 +446,96 @@ func (p *ProjectRef) AddToRepoScope(user *user.DBUser) error {
 	return nil
 }
 
+// DetachFromRepo removes the branch from the relevant repo scopes, and updates the project to not point to the repo.
+// Any values that previously defaulted to repo will have the repo value explicitly set.
+func (p *ProjectRef) DetachFromRepo(u *user.DBUser) error {
+	before, err := GetProjectSettingsEventById(p.Id)
+	if err != nil {
+		return errors.Wrap(err, "error getting before project settings event")
+	}
+
+	// remove from relevant repo scopes
+	if err = p.RemoveFromRepoScope(); err != nil {
+		return err
+	}
+
+	mergedProject, err := FindMergedProjectRef(p.Id)
+	if err != nil {
+		return errors.Wrap(err, "error finding merged project ref")
+	}
+
+	// Save repo variables that don't exist in the repo as the project variables.
+	// Wait to save merged project until we've gotten the variables.
+	mergedVars, err := FindMergedProjectVars(before.ProjectRef.Id)
+	if err != nil {
+		return errors.Wrap(err, "error finding merged project vars")
+	}
+
+	mergedProject.UseRepoSettings = false
+	mergedProject.RepoRefId = ""
+	if err = mergedProject.Upsert(); err != nil {
+		return errors.Wrap(err, "error detaching project from repo")
+	}
+
+	// catch any resulting errors so that we log before returning
+	catcher := grip.NewBasicCatcher()
+	if mergedVars != nil {
+		_, err = mergedVars.Upsert()
+		catcher.Wrap(err, "error saving merged vars")
+	}
+
+	if len(before.Subscriptions) == 0 {
+		// Save repo subscriptions as project subscriptions if none exist
+		subs, err := event.FindSubscriptionsByOwner(before.ProjectRef.RepoRefId, event.OwnerTypeProject)
+		catcher.Wrap(err, "error finding repo subscriptions")
+
+		for _, s := range subs {
+			s.ID = ""
+			s.Owner = p.Id
+			catcher.Add(s.Upsert())
+		}
+	}
+
+	// Handle each category of aliases as it's own case
+	repoAliases, err := FindAliasesForProject(before.ProjectRef.RepoRefId)
+	catcher.Wrap(err, "error finding repo aliases")
+
+	hasInternalAliases := map[string]bool{}
+	hasPatchAlias := false
+	for _, a := range before.Aliases {
+		if utility.StringSliceContains(evergreen.InternalAliases, a.Alias) {
+			hasInternalAliases[a.Alias] = true
+		} else { // if it's not an internal alias, it's a patch alias. Only add repo patch aliases if no patch aliases exist for the project.
+			hasPatchAlias = true
+		}
+	}
+	repoAliasesToCopy := []ProjectAlias{}
+	for _, internalAlias := range evergreen.InternalAliases {
+		// if the branch doesn't have the internal alias set, add any that exist for the repo
```

Comment:
I may be wrong about this, but I believe it is possible to have a mixed state where a project uses both repo level and project level aliases, and if there are two with the same name, then the project one will take precedence. If that is the case, do we want to maybe copy over all of them unless it's overwritten?

Reply:
I think based on https://github.com/evergreen-ci/evergreen/blob/3c58dd6911ad37d81323cc4d3216d6d52e030788/model/project_aliases.go#L104 that's actually technically true 🤔   but shouldn't be based on the UI design (it's supposed to be that either you default to repo, or you define your own, which is more generally consistent), so I think you've just found a bug! We may want to add some fields to ProjectRef to indicate whether or not we should be defaulting instead of doing it automatically if no aliases exist; this seems ticket-sized / will affect other things so I'm going to leave this as is but create a ticket. https://jira.mongodb.org/browse/EVG-15307

Reply:
Does this mean that if someone defaults aliases to repo, we don't want to let them define additional aliases that are specific to the project, and they will need to add it to the repo if they want to be able to use it in their project?

Reply:
discussed offline